### PR TITLE
Local url remote

### PR DESCRIPTION
**Pull request by blqn**
fix:#470

### CODE REVIEW

File: ideascube/serveradmin/catalog.py [hunk 1]
```diff
@@ -831,14 +832,18 @@ def _progress(*args):
...
-                    urlretrieve(remote.url, tmppath, reporthook=_progress)
-                except ConnectionError:
-                    print("Warning: Impossible to connect to the remote"
-                          " {remote.name}({remote.url}).\n"
-                          "Continue anyway without this remote."
-                          .format(remote=remote))
-                    continue
+                urlparsed = urlparse(remote.url)
+                if urlparsed.scheme in ['file', '']:
+                    shutil.copyfile(urlparsed.path, tmppath)
```
Comment (mgautierfr) [inline ideascube/serveradmin/catalog.py]: copyfile may raise exceptions.
You should try/except them to handle them correctly (as we do when we use urlretrieve).

**Comment by blqn** [comment]
I add the missing try catch.

File: ideascube/serveradmin/catalog.py [hunk 2]
```diff
@@ -831,14 +832,25 @@ def _progress(*args):
...
                 # TODO: Verify the download with sha256sum? Crypto signature?
-                try:
-                    urlretrieve(remote.url, tmppath, reporthook=_progress)
-                except ConnectionError:
-                    print("Warning: Impossible to connect to the remote"
-                          " {remote.name}({remote.url}).\n"
-                          "Continue anyway without this remote."
-                          .format(remote=remote))
-                    continue
+                urlparsed = urlparse(remote.url)
```
Comment (bochecha) [inline ideascube/serveradmin/catalog.py]: Minor nit: `urlparsed` and `urlparse` are very close to each other, and could cause confusion in the future.

But that's a very localized problem, `urlparsed` is used a grand total number of two times, so... ¯\_(ツ)_/¯

File: ideascube/serveradmin/catalog.py [hunk 3]
```diff
@@ -831,14 +832,25 @@ def _progress(*args):
...
+                              .format(remote=remote, error=error))
+                        continue
+                else:
+                    try:
+                        urlretrieve(remote.url, tmppath, reporthook=_progress)
+                    except ConnectionError:
+                        print("Warning: Impossible to connect to the remote"
+                            " {remote.name}({remote.url}).\n"
+                            "Continue anyway without this remote."
+                            .format(remote=remote))
```
Comment (bochecha) [inline ideascube/serveradmin/catalog.py]: You broke the proper indentation of the string here.

This is such an offense to good taste that I must immediately refuse this pull request in the most obnoxious way possible.

Just kidding, but try to be mindful about that in the future. :stuck_out_tongue_winking_eye:

File: ideascube/serveradmin/catalog.py [hunk 4]
```diff
@@ -584,37 +584,55 @@ def _progress(*args):
...
                 if self._verify_sha256(path, package.sha256sum):
                     return path
+                if urlparsed.scheme in ['file', '']:
+                    try:
+                        shutil.copyfile(urlparsed.path, path)
+                    except Exception as error:
+                        print("Warning: Impossible to copy the package file"
+                              " {package.title}({package.url}).\n{error}\n"
+                              "Continue anyway without this remote."
+                              .format(package=package, error=error))
```
Comment (bochecha) [inline ideascube/serveradmin/catalog.py]: You're copying a package, not a remote, so the message doesn't seem right.

Maybe something like `'Error copying the file: {}'.format(error)` to stay consistent with the download one?

**Comment by bochecha** [comment]
I just merged your branch after:

* rebasing it on the latest master (`git fetch --all && git rebase origin/master`)
* removing the last few instances of `fake_urlretrieve` (I added them in another pull request, so you couldn't remove them here without rebasing first)
* fixing the indentation and some wording of an error message

Compared to what you had already done this was trivial, so I figured I'd do it myself and merge it. :smiley: 

Closing this now, thanks again for your work on this, it's something we've needed for a while and it will be very useful!

**Pull request closed without merging**